pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1203,29 +1226,88 @@ func (vs *VSphere) CreateVolume(volumeOptions *vclib.VolumeOptions) (canonicalVo
 			}
 		}
 		if volumeOptions.StoragePolicyName != "" && volumeOptions.Datastore == "" {
-			datastore, err = getPbmCompatibleDatastore(ctx, dc, volumeOptions.StoragePolicyName, vs.nodeManager)
+			if len(volumeOptions.Zone) == 0 {
+				klog.V(4).Infof("Selecting a shared datastore as per the storage policy %s", volumeOptions.StoragePolicyName)
+				datastore, err = getPbmCompatibleDatastore(ctx, dc, volumeOptions.StoragePolicyName, vs.nodeManager)
+			} else {
+				// If zone is specified, first get the datastores in the zone.
+				dsList, err = getDatastoresForZone(ctx, dc, vs.nodeManager, volumeOptions.Zone)
+
+				// If unable to get any datastore, fail the operation.
+				if len(dsList) == 0 {
+					err := fmt.Errorf("Failed to find a shared datastore matching zone %s", volumeOptions.Zone)
+					klog.Error(err)
+					return "", err
+				}
+
+				klog.V(4).Infof("Specified zone : %s. Picking a datastore as per the storage policy %s among the zoned datastores : %s", volumeOptions.Zone,
+					volumeOptions.StoragePolicyName, dsList)
+				// Among the compatible datastores, select the one based on the maximum free space.
+				datastore, err = getPbmCompatibleZonedDatastore(ctx, dc, volumeOptions.StoragePolicyName, dsList)
+			}
+			klog.V(1).Infof("Datastore selected as per policy : %s", datastore)
 			if err != nil {
 				klog.Errorf("Failed to get pbm compatible datastore with storagePolicy: %s. err: %+v", volumeOptions.StoragePolicyName, err)
 				return "", err
 			}
 		} else {
-			// Since no storage policy is specified but datastore is specified, check
-			// if the given datastore is a shared datastore across all node VMs.
-			sharedDsList, err := getSharedDatastoresInK8SCluster(ctx, dc, vs.nodeManager)
-			if err != nil {
-				klog.Errorf("Failed to get shared datastore: %+v", err)
-				return "", err
-			}
-			found := false
-			for _, sharedDs := range sharedDsList {
-				if datastore == sharedDs.Info.Name {
-					found = true
-					break
+			// If zone is specified, pick the datastore in the zone with maximum free space within the zone.
+			if volumeOptions.Datastore == "" && len(volumeOptions.Zone) != 0 {
+				klog.V(4).Infof("Specified zone : %s", volumeOptions.Zone)
+				dsList, err = getDatastoresForZone(ctx, dc, vs.nodeManager, volumeOptions.Zone)
+
+				// If unable to get any datastore, fail the operation
+				if len(dsList) == 0 {
+					err := fmt.Errorf("Failed to find a shared datastore matching zone %s", volumeOptions.Zone)
+					klog.Error(err)
+					return "", err
+				}
+
+				if err != nil {
```

Comment:
Shouldn't this be ahead of the if block starting at line 1260 ?

Reply:
Yes. That's correct. Thanks for pointing out. The current sequence makes this check a no-op. Fixed it. Have created a PR for this - https://github.com/kubernetes/kubernetes/pull/74263